torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -466,8 +466,294 @@ def codegen_define(self, kernel: CppTemplateKernel) -> str:
         return result
 
 
+@register_micro_gemm(
+    *generate_gemm_config(
+        VecAVX512,
+        [(4, 64, 32), (4, 64, 64), (4, 64, 128), (4, 64, 256)],
+        input_dtype=torch.bfloat16,
+        input2_dtype=torch.int32,
+        output_dtype=torch.float,
+        compute_dtype=torch.float,
+    ),
+)
+class CppMicroGemmInt4WoQVec(CppMicroGemmFP32Vec):
+    """
+    This class generates the code for int4 WoQ micro gemm using AVX512 intrinsics.
+    """
+
+    DECLARE_KERNEL = r"""
+template <bool accum>
+inline void {{kernel_name}}(
+{%- if kernel_extra_args_declare %}
+    {{kernel_extra_args_declare}}
+{%- endif %}
+    const {{input_t}}* {{restrict_keyword}} A,
+    const {{input2_t}}* {{restrict_keyword}} B,
+    const {{input_t}}* {{restrict_keyword}} S,
+    {{output_t}}* {{restrict_keyword}} C,
+    int64_t M,
+    int64_t N,
+    int64_t K,
+    int64_t lda,
+    int64_t ldb,
+    int64_t ldc,
+    int64_t actual_N
```

Comment:
While `kernel_extra_args_declare` helps reduce duplication of code, it'd be problematic for MSVC, since the restrict keyword for it would differ from that of gcc/clang, so it can't be hard-coded like this:

<img width="650" alt="image" src="https://github.com/user-attachments/assets/58f49715-0b73-469e-9d00-dda5abe7471a">